protocol.go

```diff
@@ -0,0 +1,80 @@
+package tarantool
+
+import (
+	"fmt"
+)
+
+type ProtocolVersion uint64
+type Feature uint64
+
+type idInfo struct {
+	version  ProtocolVersion
+	features []Feature
+}
+
+const NoProtocolVersion ProtocolVersion = 0
+
+const (
+	// Streams support.
+	StreamsFeature Feature = 0
+	// Interactive transactions support.
+	TransactionsFeature Feature = 1
+	// Support of MP_ERROR objects over MessagePack.
+	ErrorExtensionFeature Feature = 2
+	// Support of watchers.
+	WatchersFeature Feature = 3
+)
+
+var featureName = map[Feature]string{
+	StreamsFeature:        "StreamsFeature",
+	TransactionsFeature:   "TransactionsFeature",
+	ErrorExtensionFeature: "ErrorExtensionFeature",
+	WatchersFeature:       "WatchersFeature",
+}
+
+// Protocol version supported by connector. Version 3
+// was introduced in Tarantool 2.10.0 and used in latest 2.10.4.
+const ClientProtocolVersion ProtocolVersion = 3
+
+// Protocol features supported by connector.
+var ClientFeatures = []Feature{
+	StreamsFeature,
+	TransactionsFeature,
+}
+
+func unsupportedFeatureError(feature Feature) error {
+	name, ok := featureName[feature]
+	if !ok {
+		return fmt.Errorf("feature with code %d is not supported", feature)
+	}
+
+	// "Feature" already is a part of the name.
+	return fmt.Errorf("%s is not supported", name)
+}
+
+func checkProtocolVersion(expected ProtocolVersion, actual ProtocolVersion) error {
+	if expected > actual {
+		return fmt.Errorf("protocol version %d is not supported", expected)
+	}
+
+	return nil
+}
+
+func checkFeatures(expectedList []Feature, actualList []Feature) error {
+	var found bool
+	// It seems that iterating over a small list is way faster
+	// than building a map: https://stackoverflow.com/a/52710077/11646599
+	for _, expected := range expectedList {
+		found = false
+		for _, actual := range actualList {
+			if expected == actual {
+				found = true
+			}
+		}
+		if !found {
+			return unsupportedFeatureError(expected)
+		}
+	}
+
+	return nil
+}
```

Comment:
1. It looks like implementation details related to `Connection`
2. You have types, you can add methods to them. To make it more general, public and tested.

Reply:
1. Moved to connection.
2. I think I cannot add method to a slice, isn't it?